Userland/Libraries/LibDSP/Track.h

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright (c) 2021, kleines Filmröllchen <malu.bertsch@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include "Clip.h"
+#include "Music.h"
+#include "Processor.h"
+#include <LibCore/Object.h>
+
+namespace LibDSP {
+
+// A track is also known as a channel and serves as a container for the audio pipeline: clips -> processors -> mixing & output
+class Track : public Core::Object {
+    C_OBJECT_ABSTRACT(Track)
+public:
+    Track(NonnullRefPtr<Transport> transport)
+        : m_transport(transport)
+    {
+    }
+    virtual ~Track()
+    {
+    }
+
+    virtual bool check_processor_chain_valid() const = 0;
+    bool add_processor(NonnullRefPtr<Processor> new_processor);
+
+    // Creates the current signal of the track by processing current note or audio data through the processing chain
+    Sample current_signal();
+
+    NonnullRefPtrVector<Processor> processor_chain() const { return m_processor_chain; }
+    const NonnullRefPtr<Transport> transport() const { return m_transport; }
+
+protected:
+    bool check_processor_chain_valid_with_initial_type(SignalType initial_type) const;
+
+    // Subclasses override to provide the base signal to the processing chain
+    virtual Signal current_clips_signal() = 0;
+
+    NonnullRefPtrVector<Processor> m_processor_chain {};
+    const NonnullRefPtr<Transport> m_transport;
+};
+
+class NoteTrack final : public Track {
+public:
+    virtual ~NoteTrack()
+    {
+    }
+
+    bool check_processor_chain_valid() const override;
+    NonnullRefPtrVector<NoteClip> clips() const { return m_clips; }
```

Comment:
This copies the whole `NonnullRefPtrVector`, is this what you want?

Reply:
No, do I do Vector<RefPtr> instead?

Reply:
`NonnullRefPtrVector<NoteClip> const&`, I guess.